# Part One

In [3]:
import dask.dataframe as dd

In [5]:
# compile csvs into dask dataframe
df = dd.read_csv('../*.csv')
df.head()

COMMENT_ID            AUTHOR  \
0  LZQPQhLyRh80UYxNuaDWhIGQYNQ96IuCg-AYWqNPjpU         Julius NM   
1  LZQPQhLyRh_C2cTtd9MvFRJedxydaVW-2sNg5Diuo4A       adam riyati   
2  LZQPQhLyRh9MSZYnf8djyk0gEF9BHDPYrrK-qCczIY8  Evgeny Murashkin   
3          z13jhp0bxqncu512g22wvzkasxmvvzjaz04   ElNino Melendez   
4          z13fwbwp1oujthgqj04chlngpvzmtt3r3dw            GsMega   

                  DATE                                            CONTENT  \
0  2013-11-07T06:20:48  Huh, anyway check out this you[tube] channel: ...   
1  2013-11-07T12:37:15  Hey guys check out my new channel and our firs...   
2  2013-11-08T17:34:21             just for test I have to say murdev.com   
3  2013-11-09T08:28:43   me shaking my sexy ass on my channel enjoy ^_^    
4  2013-11-10T16:05:38            watch?v=vtaRGgvGtWQ   Check this out .   

   CLASS  
0      1  
1      1  
2      1  
3      1  
4      1

In [6]:
# confirm we have our rows
len(df)

1956

In [7]:
# Our data is grouped by Spam (1) and Not Spam (0) in CLASS.
df.groupby('CLASS').count().compute()

COMMENT_ID  AUTHOR  DATE  CONTENT
CLASS                                   
0             951     951   951      951
1            1005    1005   760     1005

Looks about right; We have 1005 classified as spam and 951 that aren't.  Now let's look at the word check as it pertains to spam messages.

In [16]:
# convert content to lower case strings
df['LC'] = df['CONTENT'].str.lower();
df['LC'].head()

0    huh, anyway check out this you[tube] channel: ...
1    hey guys check out my new channel and our firs...
2               just for test i have to say murdev.com
3     me shaking my sexy ass on my channel enjoy ^_^ 
4              watch?v=vtarggvgtwq   check this out .
Name: LC, dtype: object

In [23]:
# classifying rows with check in the string
df['CHECK'] = df['LC'].apply(lambda i: 'check' in i, meta=('str'));
df.CHECK.head(10)

0     True
1     True
2    False
3    False
4     True
5     True
6    False
7     True
8     True
9     True
Name: CHECK, dtype: bool

In [33]:
df.groupby(['CLASS', 'CHECK']).count().compute()

COMMENT_ID  AUTHOR  DATE  CONTENT   LC
CLASS CHECK                                        
0     False         932     932   932      932  932
      True           19      19    19       19   19
1     False         544     544   466      544  544
      True          461     461   294      461  461

Above we have a confusion matrix of sorts.  As expected of non-spam entries only 19 have the word check as opposed to 461 of those classified as spam.

---
# Part Two

You've been introduced to a variety of platforms (AWS SageMaker, AWS EMR, Databricks), libraries (Numba, Dask, MapReduce, Spark), and languages (Python, SQL, Scala, Java) that can "scale up" or "scale out" for faster processing of big data.

Write a paragraph comparing some of these technology options. For example, you could describe which technology you may personally prefer to use, in what circumstances, for what reasons.

---


The options this space a pretty diverse but I think there is a pretty clear order of operations for useage. 

To begin, I prefer to start with technologies that integrate into technologies with which i'm familiar.  Dask and Numba fall into this category nicely.  By allowing simple parrallelization without infringing on my use of vanilla Python and jupyter notebooks they offer a great efficiency without much downside.

MapReduce and Spark have their place with data sets of a certain size and with operations that work well concurrently. Spark is still flexible, as it integrates well over a number of languages (Python, Scala, Java, SQL) but it has some overhead in time spent learning the tool and perhaps a slight ding when compared with Dask or Numba.

I have yet to encounter data that truly _requires_ the use of SageMaker and EMR with AWS but i'm sure they have their place.  There is the complexity associated with the dashboards and setting up clusters but once in the notebook the interface is still familiar.  Even with larger data sets I would still start by sampling a managable chunck of my data set, working with Dask/Numba, then moving into more powerful tools.  I will say, Data Bricks is really easy to use and provides enough power to get into some bigger data for free.

In short, I think I prefer to work nimbly as long as possible and save the heavy hitter for when they're truly needed. I think this flexbility would lend itself to exploration and pivoting as I find the best way to utilize the data.